added new page describing tools components

joseph-fajen opened the pull request:
Regarding the new page describing all the Tools logical components, I pushed Draft 2 (then an updated Draft 2) today (Sept. 28) for your review and feedback/comments.
We now have an interesting diagram included to look at! 
To see the updated draft rendered as an html page, please use this link:

https://plutus-apps--711.org.readthedocs.build/en/711/

Then navigate to the first page of the Explanations section--“Plutus tools in development.”
Certain comment threads need further discussion, so I embedded those directly into the updated Draft 2 page to call them out.
Biggest changes:
1. removed references to dependencies
2. removed “State” descriptions
3. The page presents all the tools as “in development.”

Thanks for your time and input!

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+`Plutus script utils <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-script-utils>`_ 
+is a utility library for helping users write Plutus scripts that are to be used on-chain. 
+Currently, this is a “kitchen sink” library containing code that was moved out of 
+plutus-ledger-api. 
```
I think kitchen sink might be better called: a variety of useful functions for on-chain operations in Plutus scripts.

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+
+`Marconi <https://github.com/input-output-hk/plutus-apps/tree/main/marconi>`_ is an 
+application for indexing useful data from the Cardano blockchain in a scalable way. 
+
```
We might want to mention that it is intended to eventually replace the chain-index

andreabedini replied:
Do we have a precise reason to say "scalable way" here? Marconi is basically a learner version of the chain-index, which makes it faster and more flexible.

joseph-fajen replied:
After checking with Konstantinos, no, "scalable way" is not a required phrase. Emphasizing flexibility is more the focus.

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
++--------------+---------------------------------------------------------------+
+| State        | In a very early alpha version. Has only a few indexers and    |
+|              | the final architecture is currently being designed.           |
++--------------+---------------------------------------------------------------+
```
I think the architecture is close to final, maybe we want to say its just not fully implemented yet? What do you think @raduom

joseph-fajen replied:
It appears that the description we have now for State is essentially accurate. But maybe we can use language that has a more affirming tone. For example: 

"State: Currently in an early alpha version. Has some indexers. Some aspects of the architecture are being designed."

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+
+`Plutus ledger <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-ledger>`_ 
+is a "kitchen sink" set of components containing data types and functions that 
+complement `cardano-ledger <https://github.com/input-output-hk/cardano-ledger>`_ 
```
I think kitchen sink here means a set of transitional types that simplify the cardano-api types. This library is being phased out in favor of the real cardano-api types?

berewt replied:
We have replaced many parts with cardano-api types, but we still simplify them a bit, as we only consider the last era. I guess the goal is to build a comprehensive, easy-to-use set of types replicating the current era of cardano-api.

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+
+4. Plutus ledger constraints
+---------------------------------
+
```
I'm confused about the difference between ledger constraints and Tx constraints. They seem to have the same description.

joseph-fajen replied:
If I'm understanding correctly, "Plutus Tx constraints" will eventually replace "Plutus ledger constraints." If so, let's move "Plutus Tx constraints" higher in the list to put more emphasis on it.

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+
+6. Plutus chain index
+---------------------------------
+
```
Soon to be deprecated in favor of Marconi

zliu41 replied:
I wouldn't even include it.

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+
+`Contract monad <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-contract>`_ 
+is a library for writing Plutus contracts. It provides the Contract Monad interface 
+for writing the off-chain part of a Plutus application that is to be interpreted 
```
A little more description of what it does: The Contract Monad provides and effect system for describing smart contracts that interact with wallets, DApps  and the blockchain.

joseph-fajen replied:
@asutherlandus I’m slightly unsure of your meaning, maybe because of a typo? Could you please say a little bit more about what you mean here just to clarify?

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+
+8. Contract monad emulator
+-----------------------------------------------------------
+
```
I think we should have the emulator adjacent with the simulator and we need to clearly distinguish between them.

joseph-fajen replied:
Noted.

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+by simulated wallets that use the contract. The component is highly dependent on the 
+Contract Monad. However, there are plans to decouple it from the Contract Monad 
+so that it can be used with other libraries.
+
```
I would eliminate the last sentence until we really know what the plan is.

joseph-fajen replied:
I've removed that last sentence.

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+
+`Plutus contract state machine <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-contract>`_ 
+is a library that provides a high-level model for writing a Plutus application based 
+on the State Machine formalism. Even though it is inefficient, it can be used to 
```
By plutus application do we mean smart contract?

joseph-fajen replied:
I believe so. Do the terms "Plutus application" and "smart contract" always mean the same thing when we talk about the Tools? If so, should we favor the term "smart contract" generally

koslambrou replied:
I like "plutus applications" (which includes on-chain + off-chain code using Plutus in Cardano) better than "smart contracts" because it's more precise. Downside is that it's not as notorious as the term "smart contract".

joseph-fajen replied:
Given all these comments, I think I should use both terms and make it clear that they are referring to the same thing.

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+is a library that provides a high-level model for writing a Plutus application based 
+on the State Machine formalism. Even though it is inefficient, it can be used to 
+model a Plutus application and to test the actual implementation.
+
```
Generally this shouldn't be used in production as the scripts are too bit to run on chain.

joseph-fajen replied:
Noted.

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+in each one. 
+
+   See also: :doc:`Testing Plutus Contracts with Contract Models <../tutorials/contract-models>`. 
+
```
There is another link to the 2nd tutorial on adversarial testing we should add here. Lets really sell this one its good stuff!

joseph-fajen replied:
@asutherlandus Are you talking about "Property-based testing of Plutus contracts"? https://plutus-apps.readthedocs.io/en/latest/plutus/tutorials/contract-testing.html

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+`PAB simulator <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-pab>`_ 
+is a simulator for the PAB that is used for testing a plutus application in a 
+simulated environment. 
+
```
We need more details. How is this different from the emulator?

asutherlandus replied:
I'm not sure I agree this is not a high priority component. It may be important when we roll out the new testing framework. @raduom what do you think?

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+`Plutus application template <https://github.com/input-output-hk/plutus-starter>`_ 
+is a Haskell template offering a simple starter project to assist dApp 
+developers in starting to write Plutus applications. 
+
```
Is this one deprecated?

koslambrou replied:
In it's current form, yes. But it's a must-have in the future because it's a way of getting entry-level plutus-application developers into the field. Do we want to remove it in the time being?

james-iohk replied:
I think people keep trying to use it still so it might be worth deprecating for now

joseph-fajen replied:
Until we have a version of this that we can genuinely promote, seems like it would be best not to mention it. I'll remove it for this version of the tools overview page.

asutherlandus commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+.. _plutus_tools_component_descriptions:
+
```
I don't see the playground should it be in here? Deprecated!

koslambrou replied:
You want to list the playground and mark it as deprecated?

joseph-fajen replied:
I think it would be better to not mention it at all if it is truly deprecated. Other thoughts?

sjoerdvisscher commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+| Dependencies | cardano-api, plutus-ledger-api, plutus-ledger,                |
+|              | plutus-ledger-constraints, plutus-chain-index-core            |
++--------------+---------------------------------------------------------------+
+| State        | Maintained but not extended. Unstable interface.              | 
```
This sounds quite negative, as if it is not intended to be used. Maybe "Maintained with a stable set of features"?

andreabedini replied:
Yeah, I was a bit confused too by the wording "Maintained but not extended. Unstable interface" Why is it unstable if we are not going to change it?

joseph-fajen replied:
I agree that it is sending mixed messages and could be confusing and potentially discouraging. I'll talk with Konstantinos about a more effective way to communicate about this issue.

joseph-fajen replied:
After discussing with Konstantinos, we decided to remove the "State" throughout this page because it's not very helpful information for readers. Instead, we are characterizing the overall Tools project as "in development." See Draft 2.

sjoerdvisscher commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+| Dependencies | plutus-contract                                               |
++--------------+---------------------------------------------------------------+
+| State        | We maintain it, but it is not production ready. It is         |
+|              | inefficient---it creates Plutus applications that have high   |
```
```suggestion
|              | inefficient---it creates Plutus applications that have too high   |
```

joseph-fajen replied:
@sjoerdvisscher I think you're making an important point about how we are describing the tools. I agree that we need to really consider how our descriptions may impact others' perceptions of the tools. Who are we targeting as our audience here? How do we want to impact them?

joseph-fajen replied:
Why would we want to talk about items for which we say "don't use it?" Why have them in the repo? Would it make more sense to get rid of it so people don't find it?

sjoerdvisscher reviewed (approved):
Nice overview. But I think that overall the state is described in too negative words. Reading this as a developer I'd skip the Plutus tools altogether, and I don't think that's the goal of this document? I think we need different levels of unstableness of interfaces: some components are quite stable but only evolve to support new features, others can possibly still completely change.

zliu41 commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+Currently, this is a “kitchen sink” library containing code that was moved out of 
+plutus-ledger-api. 
+
+It provides: 
```
I wouldn't say that it only provides these three things. How about "It provides a number of utilities including:"

joseph-fajen replied:
Noted. Thanks.

zliu41 commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+`plutus-apps <https://github.com/input-output-hk/plutus-apps>`_ repository. 
+
+For each tool listed below, we have indicated its specific location within 
+plutus-apps, provided a short description, indicated its dependencies, and given 
```
I don't think it's that useful to list the dependencies. You'll have to maintain them otherwise they can easily get out of date.

michaelpj replied:
Yes, this seems kind of crazy to me! People who care can look in the cabal files! We _really_ don't want to duplicate information like that.

joseph-fajen replied:
Great point. I'll discuss with Konstantinos.

joseph-fajen replied:
After discussing with Konstantinos, we have removed "Dependencies" throughout this page since that info is too detailed for this page, and as Ziyang and Michael pointed out, too difficult to maintain.

zliu41 reviewed (changes requested):
By reading this I feel like almost everything is either low priority, or unstable, or not maintained, or being deprecated, or more than one of these things. Could you re-order the components, and list the components that will be part of the final offering first, and then the rest of the components in a separate section?

raduom commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+---------------------------------
+
+`Plutus chain index <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-chain-index-core>`_ 
+is an application for indexing useful data from the Cardano blockchain that is to be 
```
```suggestion
is an application for indexing data from the Cardano blockchain that is to be 
```

joseph-fajen replied:
I agree with your suggestion to remove "to be" in this situation.

berewt reviewed (changes requested):
As @zliu41, I think that the status field gives a negative feeling about the tools. Removing the status and grouping them by their activity status first (active / maintenance mode / deprecated (if any)) and, for the active ones, distinguishing the ones that are ready and the ones that are under development should provide a better impression and more clarity.

This aside, the overview looks great to me.

effectfully commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+| Dependencies | plutus-ledger, plutus-ledger-api, plutus-ledger-constraints,  |
+|              | plutus-contract                                               |
++--------------+---------------------------------------------------------------+
+| State        | Maintained and worked on. Developed by QuviQ.                 | 
```
This is a little ambiguous: does "maintained and worked on" mean that it's ready for production use? I assume it does, but then I'd prefer if it was explicitly stated that the user is blessed to use this library.

joseph-fajen replied:
Great point. I'll find out more details about what is intended to be communicated here in that regard.

effectfully commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
++--------------+---------------------------------------------------------------+
+| State        | Maintained. The examples are used as test cases.              |
+|              | While they work in the plutus-contract emulator, they         |
+|              | are not guaranteed to work on the actual Cardano network.     | 
```
Again, until it's fixed I believe it's worth being transparent and warn people right away. Unless someone can quickly fix up the bug, in which case it's indeed not worth putting the warning here, but I assume it's not that easy.

joseph-fajen replied:
Noted. I'll investigate this a bit further.

effectfully reviewed (commented):
Here's my rather unhelpful review.

I agree with @zliu41 that listing dependencies doesn't have a good benefit-to-cost ratio.

I think "maintained" doesn't sound positive enough. If something is usable and has a more-or-less stable interface, let's explicitly encourage people to use it.

michaelpj commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+Plutus tools component descriptions
+=================================================
+
+The Plutus team has provided a set of “Tools”---components, libraries and 
```
I think they're not "tools" but just tools! Also I don't think "components" adds anything over "libraries and executables"

michaelpj commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+| State        | Work in progress with an unstable interface.                  |
++--------------+---------------------------------------------------------------+
+
+2. Marconi
```
can't we get automatic section numbering or somethign?

joseph-fajen replied:
I bet we can use some automatic numbering here. I'll have a look at that.

joseph-fajen replied:
And that's an interesting suggestion about `README.md` files in each package directory, etc. If not for this very next iteration, I would like to head in that direction. Seems logical!

joseph-fajen replied:
The automatic numbering doesn't work with lines separated like these headings are. It appears to only work for a series of lines that are together. I've tried it a few ways without success. But doing it manually is no big time sink and it works. 

I totally agree with your suggestion about doing README.md files. I think the first stage is to get this overview page finished. The next stage would be to create or update as many README files as needed in the Tools repo to address all the logical components referenced on this page. Once that's completed, this page could be updated to link to those README files.

michaelpj commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+| State        | In the process of being deprecated in favor of Marconi.       | 
++--------------+---------------------------------------------------------------+
+
+7. Contract monad (aka contract API, plutus contract)
```
I think the latter names are more accurate. The contract monad may be the main bit we talk about, but it's not the name of the library!

joseph-fajen replied:
Hmm -- good to know. I'll double-check on these names.

michaelpj commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+| State        | Maintained but not extended. Unstable interface.              | 
++--------------+---------------------------------------------------------------+
+
+8. Contract monad emulator
```
Okay, now I'm confused: what _is_ this document. Is it a list of libraries? evidently not, since we've now got multiple components in the same library. So what _is_ it?

koslambrou replied:
This comes of a list that I produced for @joseph-fajen. The idea was to list logical components in the current plutus-apps repository, which do not necessarily have a 1-1 relationship which the cabal projects. For example, `plutus-contract` contains both the API to write plutus applications, as well as the emulator to test plutus-applications. These two logical components could easily be split into two different cabal projects, hence why I listed them as separate _logical_ components.

Seems like that was maybe not a good idea?

If the alternative is to list _actual_ cabal projects, then the idea of adding this information in the README directly and importing it directly in the RTD seems like a good idea. I agree we don't want to duplicate the list of cabal projects where a user can simply look at the repository.

joseph-fajen replied:
@koslambrou That's helpful perspective. I think it may make sense to approach this document in stages. Maybe we can address the main comments, make some clarifications and reorganizing for stage 1. Then for stage 2, we create README files where appropriate and approach it that way.

joseph-fajen replied:
I've updated the page to make it clear that we're talking about logical components, and that a single Haskell package may contain multiple logical components.

michaelpj reviewed (commented):
Here's a suggestion: what if we put a `README.md` in each package directory, with the content here or something similar (or more!) and then included all of those into the documentation site?

james-iohk commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+instances. 
+
+PAB wraps the contracts built with plutus-contract. It is the central point of 
+contact for everything, integrating all the Cardano components. 
```
It does integrate a lot but not "all", e.g. not Marconi
Could say "integrating many Cardano components"

joseph-fajen commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+--------------------------
+
+`Marconi <https://github.com/input-output-hk/plutus-apps/tree/main/marconi>`_ is an 
+application for indexing useful data from the Cardano blockchain in a scalable way. 
```
@raduom Do you think it would be worth saying that it is an application for indexing data in a customized way from the Cardano blockchain?

joseph-fajen commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+is an application for indexing useful data from the Cardano blockchain that is to be 
+used in the Contract Monad. The main design is for indexing information related to the 
+UTxO set in order to keep the amount of indexed information proportional to the 
+UTxO set. The component is not meant to index historical information, contrary to 
```
I'll get clarification about the term "historical" here.

joseph-fajen commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+Plutus tools component descriptions
+=================================================
+
+The Plutus team has provided a set of “Tools”---components, libraries and 
```
When viewed through the rendered html pages, the three dashes become a longer dash, like an em dash versus a hyphen. Do you typically view the html rendered version on Github? I just recently learned that you can do that. In case you don't already know this, try going to the bottom of this page. The click the link "Show all checks." At the bottom of that list (you may need to scroll down in the window), select the "Details" link that appears to the right of "docs/readthedocs.org:plutus-apps — Read the Docs build succeeded!" Then you can view the html rendered version like how it is meant to be seen.

joseph-fajen replied:
Reminder that you can view the rendered html pages with this link: 
https://plutus-apps--711.org.readthedocs.build/en/711/

Also, we have removed the "Dependencies" and the "State" information from this page.

joseph-fajen commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+contains a constraints-based API that can be used to generate on-chain validation 
+functions and to build transactions by providing a list of constraints. The main 
+design is to be able to use the same constraints on-chain and off-chain in a Plutus 
+application. The off-chain part generates transactions based on types in cardano-api. 
```
@james-iohk Would that be this link? 
https://input-output-hk.github.io/cardano-node/cardano-api/lib/Cardano-Api.html

joseph-fajen replied:
I've updated the references to cardano-api to be links in multiple instances.

james-iohk replied:
Sorry @joseph-fajen I missed notification of your question, That link is fine :+1:

joseph-fajen commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,321 @@
+10. Plutus contract testing model
+------------------------------------------
+
+`Plutus contract testing model <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-contract>`_ 
```
Thanks, that's helpful!

joseph-fajen commented:
I pushed Draft 2 today (Sept. 28) for review. To see the updated draft rendered as html pages, please use this link:
--> https://plutus-apps--711.org.readthedocs.build/en/711/ 
Then navigate to the first page of the Explanations section--"Plutus tools in development." 
Certain comment threads need further discussion, so I embedded those directly into the updated Draft 2 page to call them out.

raduom commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+.. _plutus_tools_component_descriptions:
+
+Plutus tools in development 
```
On the image, the Marconi section has a 'Haskell library' underneath it, but the description says it's an application. I think it is more accurate to think of it as a library rather than an application.

joseph-fajen replied:
I changed "application" to "library."

raduom commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+|              | ``plutus-apps/plutus-chain-index``                            |
++--------------+---------------------------------------------------------------+
+
+   COMMENT: Ziyang suggests that we remove this one from the page. Does anyone have other opinions? 
```
There is no replacement for `plutus-chain-index` right now. I would like to leave it here until we get one.

koslambrou replied:
Agree

joseph-fajen replied:
Noted.

raduom commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+
+`Plutus chain index <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-chain-index-core>`_ 
+is an application for indexing data from the Cardano blockchain that is 
+used in the Contract Monad. The main design is for indexing information related to the 
```
I think this sounds a little better: `The main design goal is to keep the size of the indexed information proportional to the Utxo set.`

raduom commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+`Plutus ledger <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-ledger>`_ is a set of transitional types that simplify the cardano-api types. 
+It is intended to be a comprehensive, easy-to-use set of types that replicate the current era of `cardano-api <https://input-output-hk.github.io/cardano-node/cardano-api/lib/Cardano-Api.html>`_. 
+It currently considers only the last era. 
+Plutus ledger contains data types and functions that complement `cardano-ledger <https://github.com/input-output-hk/cardano-ledger>`_ related to Plutus. 
```
I am not sure if we should add a note here. We've been discussing using hydra's version of cardano-api helpers which AFAIK does the same thing. Does this mean that we are deprecating it. Should we add a note?

joseph-fajen replied:
@raduom @koslambrou @asutherlandus 
Unless you have already implemented hydra's version of cardano-api helpers or have firm plans to implement that right away, I would think it would be better to not add a note for the time being. Please tell me if you disagree. We can update this in the near future if needed.

raduom commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+| Location     | ``plutus-apps/plutus-ledger-constraints``                     |
++--------------+---------------------------------------------------------------+
+
+`Plutus ledger constraints <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-ledger-constraints>`_ contains a constraints-based API that can be used to generate on-chain validation functions and to build ransactions by providing a list of constraints. 
```
```suggestion
`Plutus ledger constraints <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-ledger-constraints>`_ contains a constraints-based API that can be used to generate on-chain validation functions and to build transactions by providing a list of constraints. 
```

joseph-fajen replied:
Good catch!

raduom commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
++--------------+---------------------------------------------------------------+
+
+`Plutus ledger constraints <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-ledger-constraints>`_ contains a constraints-based API that can be used to generate on-chain validation functions and to build ransactions by providing a list of constraints. 
+The main design is to be able to use the same constraints on-chain and off-chain in a plutus application. 
```
```suggestion
The main design goal is to be able to use the same constraints on-chain and off-chain in a plutus application. 
```

joseph-fajen replied:
Somehow I missed this "goal" addition but now I've added it. It will be in the merged version.

raduom reviewed (approved):
Only small, very ignorable observations and some spelling corrections.

koslambrou commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+| Location     | ``plutus-apps/plutus-contract``                               |
++--------------+---------------------------------------------------------------+
+
+`Plutus contract testing model <https://github.com/input-output-hk/plutus-apps/blob/main/plutus-contract/src/Plutus/Contract/Test/ContractModel.hs>`_ is used for testing prototype Plutus contracts with *contract models*, using the framework provided by :hsmod:`Plutus.Contract.Test.ContractModel`. 
```
testing model -> model testing

joseph-fajen replied:
Got it. Thanks for catching that.

koslambrou commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+This framework generates and runs tests on the Plutus emulator, where each test may involve a number of emulated wallets, each running a collection of Plutus contracts, all submitting transactions to an emulated blockchain. 
+Once you have defined a suitable model, then QuickCheck can generate and run many thousands of scenarios, taking the application through a wide variety of states, and checking that it behaves correctly in each one. 
+
+   See also: 
```
Not sure but:
See also: -> See the following tutorial:

joseph-fajen replied:
Good idea.

koslambrou commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+| Location     | ``plutus-apps/plutus-contract``                               |
++--------------+---------------------------------------------------------------+
+
+`Contract API <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-contract>`_ is a logical component within the Plutus Contract package, providing an effective system for describing smart contracts that interact with wallets, DApps and the blockchain. 
```
effective -> effect?

koslambrou replied:
interact with wallets, DApps and the blockchain -> interact with wallets, dApps, a chain indexer and the blockchain

joseph-fajen replied:
I've incorporated your comments. By the way, our IOG style guide says that we're supposed to say "DApps" rather than "dApps."

koslambrou replied:
Nice to know :)

koslambrou commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
++--------------+---------------------------------------------------------------+
+| PAB simulator                                                                |
```
Since we're not sure about that part given the comments below, I'd propose removing it.

joseph-fajen replied:
@koslambrou @asutherlandus @raduom I'll remove PAB simulator for now. We can always put it back in if you determine that it should be listed here.

koslambrou commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+
+For example:
+
+   * ``checkScriptContext (MustSpendAtLeast 10Ada, MustProduceOutput myOutput, …)``
+   * ``mkTx (MustSpendAtLeast 10Ada, MustProduceOutput myOutput, …)``
```
Copy paste this into the `plutus-tx-constraints` component?

koslambrou commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+For each plutus application use case, we provide test scenarios (or test cases) with and without the :hsmod:`Plutus.Contract.Test.ContractModel`.
+
+The examples are for testing and educational purposes. 
+While they work in the `plutus-contract emulator`, they are not guaranteed to work on the actual Cardano network. 
```
Add the following: [...] they are not guaranteed to work on the actual Cardano network. The main reason is that the size of the produced Plutus scripts are too big to fit in a transaction given current protocol parameters.

koslambrou commented on doc/plutus/explanations/plutus-tools-component-descriptions.rst:
```diff
@@ -0,0 +1,241 @@
+
+For each tool or logical component shown below, we have indicated its specific location within plutus-apps and provided a brief description. 
+
+.. figure:: ./plutus-tools-ecosystem.png
```
I would put the source of the image somewhere. Maybe as a comment above this line for now?

By source, I mean the link to the diagram in the PlanUML webserver.

koslambrou replied:
Also, some components are missing a description. Is that okay for now?

joseph-fajen replied:
I added a reference to the source of the diagram as a link. 
I'm not sure how to add descriptions in the diagram, but once we sort that out, we can update it and post it to the page.

joseph-fajen commented:
Draft 3 is available now. If no big issues remain, I'd like to merge this. Any objections? @zliu41 and @berewt I need you to approve before I can merge.

berewt reviewed (approved):
LGTM.
Just one thing: I was disappointed to notice that the locations in the tables that introduce each components aren't hyperlinks. Maybe we can copy paste the hyperlink we use for the tool at the beginning of each next paragraph. 

> Good point. I've made those location references into links. 
> --Joseph

joseph-fajen commented:
Pushed Draft 3 for final review. 

--> https://plutus-apps--711.org.readthedocs.build/en/711/plutus/explanations/plutus-tools-component-descriptions.html

* Incorporated comments from Draft 2. 
* Re-arranged the order of the list to correspond to the diagram. 
* The location path references in the small tables are now links. 

I had to make the table structures extra long to accommodate the long links, and I suspect this may be causing a slight hiccup in the build as the last one had a couple of errors. Although the html renders this just fine. I adjusted and am attempting another build.
